File: src/crawler.rs
use std::collections::{HashSet, VecDeque};
use std::sync::Arc;

use reqwest::header::CONTENT_TYPE;
use url::Url;

use crate::concurrent::run_ordered;
use crate::html::{LinkElement, ParsedPage, is_html_content_type, parse_html};
use crate::progress::{PhaseStart, ProgressReporter};
use crate::robots::{RobotsCache, RobotsRedirectPolicy};
use crate::sitemap::{
    MAX_COMPRESSED_BYTES, ParsedSitemap, SitemapFrontier, SitemapKind, parse_sitemap_document,
};
use crate::transport::{AllowRedirects, BodyMode, GetRequest, Transport, joined_header};
use crate::url_compat::{normalize_url, same_origin};
use crate::{AuditError, Options};

const MAX_HTML_BYTES: usize = 10 * 1024 * 1024;
const NON_HTML_DRAIN_BYTES: usize = 32 * 1024;

#[derive(Clone, Debug, Eq, PartialEq)]
pub(crate) struct CrawledPage {
    pub(crate) outside_scope: bool,
    pub(crate) page: ParsedPage,
    pub(crate) url: Url,
    pub(crate) final_url: Url,
    pub(crate) depth: i64,
    pub(crate) status_code: Option<u16>,
    pub(crate) content_type: Option<String>,
}

#[derive(Clone, Debug)]
struct QueuedPage {
    url: Url,
    key: String,
    depth: i64,
}

#[derive(Clone, Debug)]
struct CrawledQueuedPage {
    queued: QueuedPage,
    crawled: CrawledPage,
}

pub(crate) async fn crawl(
    start_url: &Url,
    transport: Transport,
    robots: Arc<RobotsCache>,
    options: Arc<Options>,
    reporter: &mut ProgressReporter,
) -> Result<Vec<CrawledPage>, AuditError> {
    let initial = queued_page(start_url, 0, options.keep_fragments);
    let mut pages = Vec::with_capacity(options.max_pages.min(128));

    if !robots.allows(&initial.url) || !options.allows_page(&initial.url) {
        reporter.start_phase(PhaseStart::Sitemaps).await?;
        return Ok(pages);
    }

    reporter.page_discovered(&initial.url).await?;

    let mut frontier = VecDeque::from([initial.clone()]);
    let mut enqueued = HashSet::from([initial.key.clone()]);
    let mut crawl_scope_url = initial.url.clone();
    let redirect_policy = RobotsRedirectPolicy::new(Arc::clone(&robots), transport.clone());

    while !frontier.is_empty() && pages.len() < options.max_pages {
        let batch_size = frontier.len().min(options.concurrency);
        let batch = frontier.drain(..batch_size).collect();
        let crawled = crawl_batch(batch, &transport, &redirect_policy, &options, reporter).await?;

        for result in &crawled {
            if result.queued.key == initial.key && !options.allows_page(&result.crawled.final_url) {
                return Err(AuditError::StartRedirectOutsideScope {
                    start: start_url.to_string(),
                    destination: result.crawled.final_url.to_string(),
                });
            }

            if result.queued.key == initial.key {
                crawl_scope_url = result.crawled.final_url.clone();
            }
        }

        for result in &crawled {
            if result.queued.depth >= i64::from(options.max_depth) {
                continue;
            }

            for link in &result.crawled.page.links {
                if enqueued.len() >= options.max_pages
                    || !matches!(link.element, LinkElement::Anchor | LinkElement::Iframe)
                    || !matches!(link.url.scheme(), "http" | "https")
                    || link.url.host().is_none()
                    || !same_origin(&link.url, &crawl_scope_url)
                {
                    continue;
                }

                let next = queued_page(&link.url, result.queued.depth + 1, options.keep_fragments);
                if enqueued.contains(&next.key)
                    || !options.allows_page(&next.url)
                    || !robots.allows(&next.url)
                {
                    continue;
                }

                enqueued.insert(next.key.clone());
                reporter.page_discovered(&next.url).await?;
                frontier.push_back(next);
            }
        }

        pages.extend(crawled.into_iter().map(|result| result.crawled));
    }

    reporter.start_phase(PhaseStart::Sitemaps).await?;

    if !options.sitemaps || options.max_depth == 0 || pages.len() >= options.max_pages {
        return Ok(pages);
    }

    let mut initial_sitemaps = robots.sitemap_urls(&crawl_scope_url);

    if initial_sitemaps.is_empty() {
        initial_sitemaps.push(
            origin_path(&crawl_scope_url, "/sitemap.xml")
                .expect("crawl scope is an HTTP URL with a host"),
        );
    }

    let mut allowed_initial = Vec::new();

    for sitemap in initial_sitemaps {
        if robots.allows(&sitemap) {
            allowed_initial.push(sitemap);
        }
    }

    let sitemap_urls = crawl_sitemaps(
        allowed_initial,
        &crawl_scope_url,
        &transport,
        &robots,
        &options,
        &mut enqueued,
        reporter,
        options.max_pages - pages.len(),
    )
    .await?;

    let mut sitemap_queue = VecDeque::with_capacity(sitemap_urls.len());

    for url in sitemap_urls {
        let queued = queued_page(&url, -1, options.keep_fragments);
        reporter.page_discovered(&queued.url).await?;
        sitemap_queue.push_back(queued);
    }

    while !sitemap_queue.is_empty() {
        let batch_size = sitemap_queue.len().min(options.concurrency);
        let batch = sitemap_queue.drain(..batch_size).collect();
        let crawled = crawl_batch(batch, &transport, &redirect_policy, &options, reporter).await?;

        pages.extend(crawled.into_iter().map(|result| result.crawled));
    }

    Ok(pages)
}

async fn crawl_batch(
    batch: Vec<QueuedPage>,
    transport: &Transport,
    redirect_policy: &RobotsRedirectPolicy,
    options: &Arc<Options>,
    reporter: &mut ProgressReporter,
) -> Result<Vec<CrawledQueuedPage>, AuditError> {
    let batch_transport = transport.clone();
    let batch_options = Arc::clone(options);
    let batch_policy = redirect_policy.clone();
    let crawled = reporter
        .with_receiver_open(run_ordered(batch, options.concurrency, move |queued| {
            let transport = batch_transport.clone();
            let options = Arc::clone(&batch_options);
            let policy = batch_policy.clone();

            async move { crawl_page(queued, &transport, &policy, &options).await }
        }))
        .await?
        .map_err(AuditError::WorkerTask)?;

    for result in &crawled {
        reporter.page_crawled(&result.queued.url).await?;
    }

    Ok(crawled)
}

async fn crawl_page(
    queued: QueuedPage,
    transport: &Transport,
    redirect_policy: &RobotsRedirectPolicy,
    options: &Options,
) -> CrawledQueuedPage {
    let mut crawled = CrawledPage {
        outside_scope: false,
        page: ParsedPage::default(),
        url: queued.url.clone(),
        final_url: queued.url.clone(),
        depth: queued.depth,
        status_code: None,
        content_type: None,
    };
    let request = GetRequest::new(
        queued.url.clone(),
        BodyMode::CollectHtml {
            max_bytes: MAX_HTML_BYTES,
            non_html_drain_bytes: NON_HTML_DRAIN_BYTES,
        },
    );

    let Ok(response) = transport.get(request, redirect_policy).await else {
        return CrawledQueuedPage { queued, crawled };
    };

    crawled.final_url = normalize_url(&response.final_url, options.keep_fragments);
    crawled.status_code = Some(response.status.as_u16());
    crawled.content_type = joined_header(&response.headers, CONTENT_TYPE);
    crawled.outside_scope = !options.allows_page(&response.final_url);

    if crawled.outside_scope {
        return CrawledQueuedPage { queued, crawled };
    }

    let content_type = crawled.content_type.clone().unwrap_or_default();

    if is_html_content_type(&content_type) {
        crawled.page = parse_html(&response.body, &response.final_url);
    }

    CrawledQueuedPage { queued, crawled }
}

#[allow(clippy::too_many_arguments)]
async fn crawl_sitemaps(
    initial: Vec<Url>,
    site_url: &Url,
    transport: &Transport,
    robots: &RobotsCache,
    options: &Options,
    excluded: &mut HashSet<String>,
    reporter: &mut ProgressReporter,
    max_pages: usize,
) -> Result<Vec<Url>, AuditError> {
    if max_pages == 0 || options.max_sitemap_documents == 0 {
        return Ok(Vec::new());
    }

    let mut frontier = SitemapFrontier::new(&initial, options.max_sitemap_documents);
    let mut pages = Vec::with_capacity(max_pages.min(256));

    while let Some(document_url) = frontier.next_document() {
        if pages.len() >= max_pages {
            break;
        }

        let fetched = reporter
            .with_receiver_open(transport.get(
                GetRequest::new(
                    document_url.clone(),
                    BodyMode::Collect {
                        max_bytes: MAX_COMPRESSED_BYTES + 1,
                    },
                ),
                &AllowRedirects,
            ))
            .await?;

        reporter.sitemap_fetched(&document_url).await?;

        let Ok(response) = fetched else {
            continue;
        };

        if !response.status.is_success() {
            continue;
        }

        let Ok(parsed) = parse_sitemap_document(&response.body) else {
            continue;
        };

        if parsed.kind == SitemapKind::Index {
            let mut documents = Vec::new();

            for candidate in parsed.urls {
                if robots.allows(&candidate) {
                    documents.push(candidate);
                }
            }

            frontier.extend_from(&ParsedSitemap {
                kind: SitemapKind::Index,
                urls: documents,
            });

            continue;
        }

        for candidate in parsed.urls {
            if pages.len() >= max_pages
                || !same_origin(&candidate, site_url)
                || !options.allows_page(&candidate)
                || !robots.allows(&candidate)
            {
                continue;
            }

            let normalized = normalize_url(&candidate, options.keep_fragments);

            if excluded.insert(normalized.to_string()) {
                pages.push(normalized);
            }
        }
    }

    Ok(pages)
}

fn queued_page(input: &Url, depth: i64, keep_fragments: bool) -> QueuedPage {
    let url = normalize_url(input, keep_fragments);

    QueuedPage {
        key: url.to_string(),
        url,
        depth,
    }
}

fn origin_path(input: &Url, path: &str) -> Option<Url> {
    let mut origin = input.clone();

    origin.set_username("").ok()?;
    origin.set_password(None).ok()?;
    origin.set_path(path);
    origin.set_query(None);
    origin.set_fragment(None);

    Some(origin)
}

File: internal/crawler/crawler.go
// Package crawler discovers and fetches pages in deterministic breadth-first
// order.
package crawler

import (
	"context"
	"errors"
	"io"
	"net/url"
	"strings"
	"sync"

	"github.com/nelsonlaidev/scoutly/internal/fetcher"
	"github.com/nelsonlaidev/scoutly/internal/page"
	"github.com/nelsonlaidev/scoutly/internal/ptrutil"
	"github.com/nelsonlaidev/scoutly/internal/sitemap"
	"github.com/nelsonlaidev/scoutly/internal/urlutil"
)

const maxHTMLBytes = 10 * 1024 * 1024

var (
	ErrNilContext          = errors.New("crawler: nil context")
	ErrNilStartURL         = errors.New("crawler: start URL is nil")
	ErrNilFetcher          = errors.New("crawler: fetcher is nil")
	ErrMaxPagesNotPositive = errors.New("crawler: max pages must be positive")
	ErrNegativeMaxDepth    = errors.New("crawler: max depth must not be negative")
	ErrInvalidConcurrency  = errors.New("crawler: concurrency must be positive")
)

type Options struct {
	Allowed               func(*url.URL) bool
	SitemapURLs           []*url.URL
	SitemapURLsFor        func(*url.URL) []*url.URL
	ScanSitemaps          bool
	MaxDepth              int
	MaxPages              int
	MaxSitemapDocuments   int
	KeepFragments         bool
	Concurrency           int
	OnPageDiscovered      func(*url.URL) error
	OnPageCrawled         func(*url.URL) error
	OnSitemapPhaseStarted func() error
	OnSitemapFetched      func(*url.URL) error
	RedirectPolicy        fetcher.RedirectPolicy
}

type CrawledPage struct {
	Page        page.Page
	URL         *url.URL
	FinalURL    *url.URL
	Depth       int
	StatusCode  *int
	ContentType *string
}

type queuedPage struct {
	URL   *url.URL
	Key   string
	Depth int
}

type crawledQueuedPage struct {
	Queued queuedPage
	Page   CrawledPage
}

func Crawl(
	ctx context.Context,
	startURL *url.URL,
	httpFetcher *fetcher.Fetcher,
	options Options,
) ([]CrawledPage, error) {
	if ctx == nil {
		return nil, ErrNilContext
	}
	if startURL == nil {
		return nil, ErrNilStartURL
	}
	if httpFetcher == nil {
		return nil, ErrNilFetcher
	}
	if options.MaxPages <= 0 {
		return nil, ErrMaxPagesNotPositive
	}
	if options.MaxDepth < 0 {
		return nil, ErrNegativeMaxDepth
	}
	if options.Concurrency <= 0 {
		return nil, ErrInvalidConcurrency
	}
	allowed := options.Allowed
	if allowed == nil {
		allowed = func(*url.URL) bool { return true }
	}

	initial := createQueuedPage(startURL, 0, options.KeepFragments)
	pages := make([]CrawledPage, 0, min(options.MaxPages, 128))
	if !allowed(initial.URL) {
		if options.OnSitemapPhaseStarted != nil {
			if err := options.OnSitemapPhaseStarted(); err != nil {
				return nil, err
			}
		}
		return pages, nil
	}

	queue := []queuedPage{initial}
	enqueued := map[string]struct{}{initial.Key: {}}
	crawlScopeURL := ptrutil.Clone(initial.URL)
	if options.OnPageDiscovered != nil {
		if err := options.OnPageDiscovered(ptrutil.Clone(initial.URL)); err != nil {
			return nil, err
		}
	}

	for len(queue) > 0 && len(pages) < options.MaxPages {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		batchSize := min(len(queue), options.Concurrency, options.MaxPages-len(pages))
		batch := append([]queuedPage(nil), queue[:batchSize]...)
		queue = queue[batchSize:]

		crawled, err := crawlBatch(ctx, batch, httpFetcher, options)
		if err != nil {
			return nil, err
		}
		for _, result := range crawled {
			pages = append(pages, result.Page)
			if result.Queued.Key == initial.Key && result.Page.FinalURL != nil {
				crawlScopeURL = ptrutil.Clone(result.Page.FinalURL)
			}
		}
		for _, result := range crawled {
			if result.Queued.Depth >= options.MaxDepth {
				continue
			}

			for _, link := range result.Page.Page.Links {
				if len(pages)+len(queue) >= options.MaxPages {
					break
				}
				if !isCrawlable(link, crawlScopeURL) {
					continue
				}
				next := createQueuedPage(link.URL, result.Queued.Depth+1, options.KeepFragments)
				if _, exists := enqueued[next.Key]; exists || !allowed(next.URL) {
					continue
				}

				enqueued[next.Key] = struct{}{}
				queue = append(queue, next)
				if options.OnPageDiscovered != nil {
					if err := options.OnPageDiscovered(ptrutil.Clone(next.URL)); err != nil {
						return nil, err
					}
				}
			}
		}
	}

	if options.OnSitemapPhaseStarted != nil {
		if err := options.OnSitemapPhaseStarted(); err != nil {
			return nil, err
		}
	}
	if !options.ScanSitemaps || options.MaxDepth == 0 || len(pages) >= options.MaxPages {
		return pages, nil
	}

	initialSitemaps := cloneURLs(options.SitemapURLs)
	if options.SitemapURLsFor != nil {
		initialSitemaps = cloneURLs(options.SitemapURLsFor(ptrutil.Clone(crawlScopeURL)))
	}
	if len(initialSitemaps) == 0 {
		initialSitemaps = []*url.URL{{
			Scheme: crawlScopeURL.Scheme,
			Host:   crawlScopeURL.Host,
			Path:   "/sitemap.xml",
		}}
	}

	discoveryContext, cancelDiscovery := context.WithCancelCause(ctx)
	defer cancelDiscovery(nil)

	var callbackMu sync.Mutex
	var callbackErr error
	sitemapURLs, err := sitemap.Crawl(discoveryContext, initialSitemaps, httpFetcher, sitemap.Options{
		Allow:            allowed,
		Scheme:           crawlScopeURL.Scheme,
		Host:             crawlScopeURL.Host,
		KeepFragments:    options.KeepFragments,
		ExcludedPageKeys: enqueued,
		MaxDocuments:     options.MaxSitemapDocuments,
		MaxPageURLs:      options.MaxPages - len(pages),
		OnDocumentFetched: func(currentURL *url.URL) {
			callbackMu.Lock()
			defer callbackMu.Unlock()
			if callbackErr != nil {
				return
			}
			if options.OnSitemapFetched != nil {
				callbackErr = options.OnSitemapFetched(ptrutil.Clone(currentURL))
			}
			if callbackErr != nil {
				cancelDiscovery(callbackErr)
			}
		},
	})
	callbackMu.Lock()
	progressErr := callbackErr
	callbackMu.Unlock()
	if progressErr != nil {
		return nil, progressErr
	}
	if err != nil {
		return nil, err
	}

	sitemapQueue := make([]queuedPage, 0, len(sitemapURLs))
	for _, pageURL := range sitemapURLs {
		queued := createQueuedPage(pageURL, -1, options.KeepFragments)
		if _, exists := enqueued[queued.Key]; exists {
			continue
		}
		enqueued[queued.Key] = struct{}{}
		sitemapQueue = append(sitemapQueue, queued)
		if options.OnPageDiscovered != nil {
			if err := options.OnPageDiscovered(ptrutil.Clone(queued.URL)); err != nil {
				return nil, err
			}
		}
	}

	for len(sitemapQueue) > 0 && len(pages) < options.MaxPages {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		batchSize := min(len(sitemapQueue), options.Concurrency, options.MaxPages-len(pages))
		batch := append([]queuedPage(nil), sitemapQueue[:batchSize]...)
		sitemapQueue = sitemapQueue[batchSize:]

		crawled, err := crawlBatch(ctx, batch, httpFetcher, options)
		if err != nil {
			return nil, err
		}
		for _, result := range crawled {
			pages = append(pages, result.Page)
		}
	}

	return pages, nil
}

func crawlBatch(
	ctx context.Context,
	batch []queuedPage,
	httpFetcher *fetcher.Fetcher,
	options Options,
) ([]crawledQueuedPage, error) {
	batchContext, cancel := context.WithCancelCause(ctx)
	defer cancel(nil)

	results := make([]crawledQueuedPage, len(batch))
	var waitGroup sync.WaitGroup
	for index, queued := range batch {
		waitGroup.Go(func() {
			if batchContext.Err() != nil {
				return
			}

			crawled, err := crawlPage(batchContext, queued, httpFetcher, options)
			if err != nil {
				cancel(err)
				return
			}
			results[index] = crawledQueuedPage{Queued: queued, Page: crawled}
			if options.OnPageCrawled != nil {
				if err := options.OnPageCrawled(ptrutil.Clone(queued.URL)); err != nil {
					cancel(err)
				}
			}
		})
	}
	waitGroup.Wait()

	if cause := context.Cause(batchContext); cause != nil {
		return nil, cause
	}
	return results, nil
}

func crawlPage(
	ctx context.Context,
	queued queuedPage,
	httpFetcher *fetcher.Fetcher,
	options Options,
) (CrawledPage, error) {
	result := CrawledPage{
		Page:     page.New(),
		URL:      ptrutil.Clone(queued.URL),
		FinalURL: ptrutil.Clone(queued.URL),
		Depth:    queued.Depth,
	}

	response, err := httpFetcher.Fetch(ctx, queued.URL, options.RedirectPolicy)
	if err != nil {
		if ctxErr := ctx.Err(); ctxErr != nil {
			return CrawledPage{}, ctxErr
		}
		return result, nil
	}
	if response == nil || response.Body == nil {
		return result, nil
	}
	defer func() {
		_ = response.Body.Close()
	}()

	baseURL := queued.URL
	if response.Request != nil && response.Request.URL != nil {
		baseURL = response.Request.URL
		result.FinalURL = urlutil.Normalize(response.Request.URL, options.KeepFragments)
	}

	statusCode := response.StatusCode
	result.StatusCode = &statusCode
	contentTypeValue, hasContentType := response.Header["Content-Type"]
	contentType := strings.Join(contentTypeValue, ", ")
	if hasContentType {
		result.ContentType = &contentType
	}

	if !page.IsHTMLContentType(contentType) {
		_, _ = io.Copy(io.Discard, io.LimitReader(response.Body, 32*1024))
		result.Page.ContentType = contentType
		return result, nil
	}

	limited := &io.LimitedReader{R: response.Body, N: maxHTMLBytes + 1}
	parsed, err := page.Parse(limited, baseURL)
	if err != nil || limited.N <= 0 {
		if ctxErr := ctx.Err(); ctxErr != nil {
			return CrawledPage{}, ctxErr
		}
		result.StatusCode = nil
		result.ContentType = nil
		result.Page = page.New()
		return result, nil
	}
	parsed.ContentType = contentType
	result.Page = parsed
	return result, nil
}

func createQueuedPage(input *url.URL, depth int, keepFragments bool) queuedPage {
	normalized := urlutil.Normalize(input, keepFragments)
	return queuedPage{URL: normalized, Key: normalized.String(), Depth: depth}
}

func isCrawlable(link page.Link, scopeURL *url.URL) bool {
	if link.URL == nil {
		return false
	}
	if link.Element != page.LinkElementAnchor && link.Element != page.LinkElementIframe {
		return false
	}
	if !urlutil.IsHTTPWithHost(link.URL) {
		return false
	}
	return urlutil.Origin(link.URL) == urlutil.Origin(scopeURL)
}

func cloneURLs(input []*url.URL) []*url.URL {
	result := make([]*url.URL, 0, len(input))
	for _, item := range input {
		result = append(result, ptrutil.Clone(item))
	}
	return result
}
